clippy_lints/src/append_instead_of_extend.rs

```diff
@@ -0,0 +1,61 @@
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+// use rustc_ast::ast::*;
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::is_type_diagnostic_item;
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// **What it does:**
+    ///
+    /// **Why is this bad?**
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // example code where clippy issues a warning
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// ```
+    pub APPEND_INSTEAD_OF_EXTEND,
+    perf,
+    "default lint description"
+}
+
+declare_lint_pass!(AppendInsteadOfExtend => [APPEND_INSTEAD_OF_EXTEND]);
+
+impl<'tcx> LateLintPass<'tcx> for AppendInsteadOfExtend {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if is_vec_an_append(cx, expr) {
+            span_lint_and_help(
+                cx,
+                APPEND_INSTEAD_OF_EXTEND,
+                expr.span,
+                "use of `vec.extend(other_vec.drain(..))`",
+                None,
+                "consider using `vec![].append(&mut vec![]);` instead",
+            );
+        }
+    }
+}
+
+fn is_vec_an_append<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> bool {
+    if_chain! {
+        if let ExprKind::MethodCall(method_name, _, exprs, _) = expr.kind;
+        if method_name.ident.as_str() == "extend";
+        let ty = cx.typeck_results().expr_ty(&exprs[0]).peel_refs();
+        if is_type_diagnostic_item(cx, ty, sym::vec_type);
+        if let ExprKind::MethodCall(method_name2, _, _exprs2, _) = &exprs[1].kind;
+        if method_name2.ident.as_str() == "drain";
+        then {
```

Comment:
Okay, you've now added the type check for the first object. We now need the exact same check for the second struct as well to make sure that it's also a `std::vec::Vec<>`.

This as well as the range check will add new if checks. It is often useful to group these checks and annotate them with a comment, like this:
```rust
// Validate target object 
if let ExprKind::MethodCall(method_name, _, exprs, _) = expr.kind;
if method_name.ident.as_str() == "extend";
let ty = cx.typeck_results().expr_ty(&exprs[0]).peel_refs();
if is_type_diagnostic_item(cx, ty, sym::vec_type);
// Check source object
if let ExprKind::MethodCall(src_method, _, src_args, _) = &exprs[1].kind;
if src_method.ident.as_str() == "drain";
let src_ty = cx.typeck_results().expr_ty(&src_args[0]).peel_refs();
if is_type_diagnostic_item(cx, src_ty, sym::vec_type);
// Check drain range
...
then {
```

---

Now about the range check:

Clippy has a utility function that tries to parse a given expression into range. You can find the function here: [`use clippy_utils::higher::range;`](https://github.com/rust-lang/rust-clippy/blob/35e8be7407198565c434b69c5b9f85c71f156539/clippy_utils/src/higher.rs#L51). You can find a usage example from a different lint [here](https://github.com/rust-lang/rust-clippy/blob/12fce557669a0de230399cf8e6eee4f5307bf87b/clippy_lints/src/methods/iter_next_slice.rs#L31-L32)

Reply:
Thanks for the updates. Added the check and test are passing. I used a slightly different method though to check the range `is_type_lang_item(cx, ty, LangItem::RangeFull)` .

Reply:
The use of `LangItem::RangeFull` is a nice solution and I've learned something new. Feel free to resolve this comment @Valentine-Mario :)